doc/sphinx/proof-engine/tactics.rst

```diff
@@ -1588,17 +1588,26 @@ Controlling the proof flow
       While :tacn:`pose proof` expects that no existential variables are generated by
       the tactic, :tacn:`epose proof` removes this constraint.
 
-.. tacv:: specialize (@ident {* @term}) {? as @simple_intropattern}
-          specialize @ident with @bindings {? as @simple_intropattern}
-   :name: specialize; _
+.. tacn:: specialize @one_term_with_bindings {? @as_ipat }
+
+   .. todo PR: Where is @ident now?  Need to adapt
 
    This tactic works on local hypothesis :n:`@ident`. The
    premises of this hypothesis (either universal quantifications or
-   non-dependent implications) are instantiated by concrete terms coming either
-   from arguments :n:`{* @term}` or from :ref:`bindings`.
-   In the first form the application to :n:`{* @term}`  can be partial. The
-   first form is equivalent to :n:`assert (@ident := @ident {* @term})`. In the
-   second form, instantiation elements can also be partial. In this case the
+   non-dependent implications) are instantiated by concrete terms from
+   :n:`@one_term_with_bindings`.
+
+   .. todo PR:
+      1) what is "non-dependent implication" above?
+      2) In this little example, what exactly is (H x)? It looks almost like
+         a function but H doesn't look like a function.  Is this just something
+         wierd about specialize or have I missed some basic point?
+
+      Goal forall x: nat, (forall n, n >= 0) -> True.
+      intros.  (* H is forall n : nat, n >= 0 *)
+      specialize (H x). (* H is x >= 0 *)
```

Comment:
You have missed some basic point. In this context, `H` is a hypothesis of `forall n : nat, n >= 0`, therefore it *is* a function that for any `n` of type `nat` returns a proof that `n >= 0`. Therefore, it can be applied to `x` (which is of type `nat` in this context).

The only thing that is special about `specialize` here (compared to e.g., `pose proof`), is that if you do not provide an `as` clause, then Coq will reuse the name `H` for the specialized hypothesis (exactly as if `specialize (H x) as H` or `pose proof (H x) as H`).

Reply:
I expect a function would look something like `(fun (x : nat) (H : forall n : nat, n >= 0) => ?Goal)` rather than a `forall`.  Is it that H really is a function but the UI only shows the type of H?  Or is H exactly the displayed type, which gets mapped/converted back into a function?  Whichever it is, I don't think that is covered anywhere in the doc.

Reply:
Coq prints : `H : forall...`, which definitely means that the *type* of H is `forall...`. Otherwise you would have `H := forall...`.
However sometimes we say that the hypothesis `H` *is* `forall...` because that is how mathematicians talk about hypothesis, but this is misleading.

Reply:
Hi Pierre,

Thanks for offering to help.  Let's take things one at a time:

`(H x)` in the specialize tactic on line 1608 looks like the application of a function (including supporting partial application).  If that's a correct way to think about it, I'd like to say something clear and strong about the correspondence between a function and its associated type (a forall).  So far I've not learned how to compute the type of a term--whether it's not in the doc at all or if it's there but not explained in an accessible form (e.g. it's in the rather dense [CIC chapter](https://coq.inria.fr/distrib/current/refman/language/cic.html)).  Fully explaining how to compute types is beyond the scope of this PR.

Or I could just make a weaker statement in this tactic noting the correspondence.  The context shows the type of H.  Every term has a type, but IIUC not every type has a corresponding term.  e.g. if H is not provable, then there is no associated term.  Even if H is provable, Coq probably doesn't know the what the term is (imagine H: <Fermat's Last Theorem>).

Reply:
I suspect the `(H x)` is a merely a convenient notation since Coq (generally?) doesn't know the specific function whose type is shown for H.

Reply:
Indeed, when `H: T` appears in a proof context, it means that `H` is a *hypothesis* so it does NOT say there is a closed term with type `T`. Instead it says that we must build a proof *under the assumption* that there is a proof of `T`, that we shall name `H`. If the type `T` happens to be a product (dependent or not), then the proof may contain applications of the form `H x`.
I think this is explained in the early chapters explaining the concepts of interactive proofs, we don't really need to come back to this in the chapters on tactics IMHO.

Reply:
There's something unexplained going on that's not explained in the doc.  I'd like to add the explanation in an appropriate section.  `(f x)` indicates applying a function f to an argument x.  From its type, `H` must be a function, but an unknown one, so how can you apply it in `(H 1)`?  Coq somehow treats the type `forall n m ...` as a function definition, which deserves an explanation.

```
Goal (forall n m, n >= m) -> True.
intros.   (* H : forall n m : nat, n >= m *)
let x := type of (H 1) in idtac x.  (* (forall m : nat, 1 >= m) *)
```

It seems to work the same way for theorems, but they don't look like functions in the `Print`.  That should be explained, too:

```
Theorem F: forall n m, n >= m.
Admitted.
Goal True.
let y := type of (F 1) in idtac y.  (* (forall m : nat, 1 >= m) *)
Print F.
(*
  *** [ F : forall n m : nat, n >= m ]

  Arguments F (_ _)%nat_scope
*)
```

Can you explain these enough so I can describe how they work?

Reply:
I see what you mean but this is old plain higher order programing : for instance, take the `List.map` function of ocaml:

```ocaml
let map (f:'a ->'b) (l:'a list): 'b list =
  match l with
  | [] -> []
  | e::l' -> f e :: map f l'
```

`f` is only a variable here and we apply if to `e`, and the type of subterm `f e` is `'b`. Do you really think we need to explain that here? I mean it is really really basic functional programing.

Reply:
When doing interactive proof we are manipulating the typing environment in the same way the typing algorithm does: in the example above when the typing algorithm reaches `f e` there is a typing environment, which includes parameters `f` and `l`, but also pattern variables `e` and `l'` that were added while diving into the term. The difference in interactive proof is that we build the term *while* typing it.
Explaining this as a preamble could be nice for users coming from outside functional programing community. But this should be done as separate section or chapter.

Reply:
Illustrating this with the map function defined interactively:
```coq

Definition map {A B:Type}: forall  (f:A -> B) (l:list A) , list B.
  fix map 2.
  intros.
  case l.
  - exact nil.
  - intros e l'.
    exact (cons (f e) (map f l')).
Defined.
```

Reply:
I hope this example also shows that there is absolutely no difference between applying a theorem, a function or a hypothesis. There is a difference in the ways these will be used afterward though: lemmas and hypothesis are rarely (never for hypothesis) used for computing, whereas functions may be called and reduced on concrete parameters.

Reply:
Also: `A -> B` is *really* just a syntactic suger for `forall x:A, B` when `B` does not mention `x`. Just type `Check (forall x:nat, bool)` in Coq to see that coq prints: `nat -> bool: Set`. This is explained [very early](https://coq.inria.fr/distrib/current/refman/language/core/assumptions.html?highlight=forall#functions-fun-and-function-types-forall) in the documentation.

Reply:
I'm well aware of these:

> `f` is only a variable here and we apply if to `e`, and the type of subterm `f e` is `'b`

> `A -> B` is *really* just a syntactic sugar for `forall x:A, B` when `B` does not mention `x`. 

This suggests my point:

> ... we build the term *while* typing it.

The evaluation process is not fully described.  H (a hypothesis) and F (a theorem) are handled differently from, say Nat.add (a fixpoint).  I think it works like this: For `(H 1)`, the type of a hypothesis H is known but its value is never known/generated/used by Coq.  Since the type of H is a dependent product, Coq infers the new type (removing the first dependent premise `n` in the `forall` ).  `let x := type of (H 1) in ...` is valid because the type is known, but `let x := (H 1) in ...` is not because the value is unknown.

Even though the value (proof) of a proven theorem is known, it appears that `let x := (F 1) in ...` is never valid, probably because it's not a useful thing to do.  Right?

In contrast, `let x := (Nat.add 1) in ...` gives you a new function.

Can you always get a value for any variable that isn't a hypothesis or a theorem?

> Explaining this as a preamble could be nice for users coming from outside functional programing community.

I think "functional programming" has no fixed meaning; there's a set of features that are associated with the term, but each language has its own definition of the term.  So good to explain.

If this seems on-track, I'll draft a proper description to add to the doc.

Reply:
Also, if the hypothesis only has nondependent premises, then `(H <something>)` is never valid?  E.g. something like this won't work even with tweaks:

```
Goal (0=0 -> 1=1) -> False.
intros.    (* H : 0 = 0 -> 1 = 1 *)
let x := type of (H (0=0)) in idtac x.
(*
  In environment
  H : 0 = 0 -> 1 = 1
  The term "0 = 0" has type "Prop"
  while it is expected to have type "0 = 0".
*)
```

Reply:
Sorry but I think there is an important aspect of coq that you misunderstand.

> The evaluation process is not fully described.

Because it does not need to be described: we build a *proof*, i.e. a term having a certain type, we don't compute with it. *After* the term is built, we may try to compute with it but it is another story.

> H (a hypothesis) and F (a theorem) are handled differently from, say, Nat.add (a fixpoint).

As far as typing is concerned: NO DIFFERENCE except for conversion: `H` does not unfold whereas theorems (even if opaque), whereas `F` and `Nat.add` do. So the only difference is between `H` and `F`, or between `H` and `Nat.add`.

> I think it works like this: For (H 1), the type of a H is known but its value is never known/generated/used by Coq.

`H` is its own value, that's all. It *is* used, as a variable.

> Since the type of H is a dependent product, Coq infers the new type (removing the first dependent premise n in the forall ).

? This has nothing to do with the product being dependent or not. In both cases `H` needs something having the *type* appearing in the left member of the product.

> let x := type of (H 1) in ... is valid because the type is known,

Yes, but are you aware that `x` has type `Prop` here?  And are you aware that the `let` here is at Ltac level here ? 

```coq
Goal (forall n:nat, n=n) -> 1=1.
intros H. 
let x := type of (H 1) in
idtac x. (* 1 = 1*)
let x := type of (H 1) in
let t := type of x in
idtac t. (* Prop *)
```

> but let x := (H 1) in ... is not because the value is unknown.

?? What? Of course it is valid, but since you are using Ltac-let you need to use `constr:` in front of the term otherwise `(H 1)` is taken as an tactic application. But it is perfectly valid, and `x` is "something of type `(H 1)`":

```coq
Goal (forall n:nat, n=n) -> 1=1.
intros H. 
let h := constr:(H 1) in exact h.
```
Using the constr-let:
```coq
Goal (forall n:nat, n=n) -> 1=1.
intros H. 
exact (let x := H 1 in x).
```

> Even though the value (proof) of a proven theorem is known, it appears that let x := (F 1) in ... is never valid, probably because it's not a useful thing to do. Right?

Wrong, very wrong. See examples above and below.

Some "names" have a _definition associated_ with them, while others don't. There are treated exactly the same by the typer except conversion. Period. Don't overthink it, there is no trick here.

`let x := t in u`, means that in `u` the _definition associated_ with name `x` is `t`. Period. No trick. Here `t` is `(F 1)`, so be it, `(F 1)` is a perfectly valid term in an environment where `F` is of type `(nat -> something)`.

> In contrast, let x := (Nat.add 1) in ... gives you a new function.

> Can you always get a value for (V value) when V doesn't refer to a hypothesis or a theorem?

Define "value" please: Since everything is a term in coq (types are terms too), "value" must be some kind of term:

- any term? `(H 1)` is a term.
- a term that is not a type? `(H 1)` is a type.
- a term in normal form? `(H (S 0))` is a term in normal form.
- a ground term (no more names in it) ? `(H 1)` is not ground.


>> Explaining this as a preamble could be nice for users coming from outside functional programing community.

> I think "functional programming" has no fixed meaning; there's a set of features that are associated with the term, but each language has its own definition of the term. Coq, OCaml, Java all considered functional, but different in many ways. So good to explain.

Right. I should have said regular typed lambda-calculus.

Actually I am starting to think you are indeed taking coq for a programming language: in ocaml `42` is meant to represent a binary integer in computer's memory once compiled and executed. In Coq (and lambda-calculus) it is **not**. It represents only itself: a term of the form `S ... (S 0)...)`. Nothing else.

> If this seems on-track, I'll draft a proper description to add to the doc.

It is not. What you are trying to explain does not exists afaiu.

> Also, if the hypothesis only has nondependent premises, then (H <something>) is never valid? E.g. something like this won't work even with tweaks:

> Goal (0=0 -> 1=1) -> False. intros. (* H : 0 = 0 -> 1 = 1 *) let x := type of (H (0=0)) in idtac x. (* In environment H : 0 = 0 -> 1 = 1 The term "0 = 0" has type "Prop" while it is expected to have type "0 = 0". *)

??? this is a standard type error, don't over-interpret it, you just made a mistake: The argument of `H` must be of TYPE `(0=0)`, not be `(0=0)` itself. If you do `H (eq_refl 0)` it is perfectly valid (and may be useful of course).

```coq
Goal (0=0 -> 1=1) -> False.
intros.    (* H : 0 = 0 -> 1 = 1 *)
let x := type of (H (eq_refl 0)) in idtac x.
```

I am sorry but I think you should not try to write in this section until you understand what you are getting wrong here.

Reply:
> I am sorry but I think you should not try to write in this section until you understand what you are getting wrong here.

That's the point of our discussion.  I'm asking about points that aren't well covered in the documentation.  I iteratively guess (often imperfectly) how things work and eventually come up with an accurate description.

Let's focus on this one point:

> > I think it works like this: For (H 1), the type of a H is known but its value is never known/generated/used by Coq.

> `H` is its own value, that's all. It *is* used, as a variable.

That doesn't make sense to me.  Consider:

```
Goal (forall n m, n >= m) -> True.
intros.  (* H : forall n m : nat, n >= m *)
let x := type of (H 1) in idtac x.
```

`(H 1)` represents the application of a function, however you seem to say that the value of `H` is `forall n m : nat, n >= m`, which is not a function AFAICT.  How does that make any sense?

--------------
Let me respond briefly to a few other comments (and come back to others a bit later):

> > The evaluation process is not fully described.

> Because it does not need to be described: we build a *proof*, i.e. a term having a certain type, we don't compute with it.

By "evaluation process" I meant whatever processing Coq does with a term, such as to check its validity.  I was not referring specifically to evaluating function applications as computations.

> Actually I am starting to think you are indeed taking coq for a programming language: in ocaml `42` is meant to represent a binary integer in computer's memory once compiled and executed. In Coq (and lambda-calculus) it is **not**. It represents only itself: a term of the form `S ... (S 0)...)`. Nothing else.

Of course not.  Everyone learns the representation of Nat at the outset of learning Coq.

Reply:
> That doesn't make sense to me. Consider:
> 
> ```
> Goal (forall n m, n >= m) -> True.
> intros.  (* H : forall n m : nat, n >= m *)
> let x := type of (H 1) in idtac x.
> ```
> 
> `(H 1)` represents the application of a function

Yes.

> however you seem to say that the value of `H` is `forall n m : nat, n >= m`, 

Again, what makes you think that?? When it is written `H : XXX` it means that `H` HAS TYPE `XXX`. It does not say anything about its value. Here `H`  HAS  TYPE  `forall n m : nat, n >= m`. `H` has no value, only a type: it is *assumed* instead of being `defined`. Whereas *constants* have both a type AND a *definition* (that you may call a value but it is misleading imho). 
The **TYPE** tells us that `H` can be applied to 2 arguments of type `nat`. So we know *by typing* that `H 1` is well typed.

> By "evaluation process" I meant whatever processing Coq does with a term, such as to check its validity. I was not referring specifically to evaluating function applications as computations.

Well, evaluation and typing are two very different things. If you want to write the documentation you should start by never naming one by the other imho. Otherwise you will confuse the reader.

- Checking the validity = typing.
- evaluation = reduction 

> > Actually I am starting to think you are indeed taking coq for a programming language: in ocaml `42` is meant to represent a binary integer in computer's memory once compiled and executed. In Coq (and lambda-calculus) it is **not**. It represents only itself: a term of the form `S ... (S 0)...)`. Nothing else.
> Of course not. Everyone learns the representation of Nat at the outset of learning Coq.

OK but you seem obsessed with "values":

> The evaluation process is not fully described. ...
> I think it works like this: For (H 1), the type of a H is known but its value is never known/generated/used by Coq. ...
> let x := type of (H 1) in ... is valid because the type is known, but let x := (H 1) in ... is not because the value is unknown.

But the whole proof mode is a type-driven term construction tool. It does not care about "values", only types. Coq's language happens to be dependently types so the typer needs to perform conversion from time to time. Which implies to use *reduction*. The reduction is the only moment the typer looks at the definitions associated with names. Otherwise looking at the *type* associated with names is enough, and there is no (*no*) difference netween variables and constants.

Reply:
The current description of function application is [here](https://coq.inria.fr/distrib/current/refman/language/core/assumptions.html#function-application).  `H` is not a function and the type of H (`forall n m : nat, n >= m`) is not a function, so how can we briefly and concretely explain, somewhere, the interpretation of `(H 1)`?  Should that be in the description of function application or elsewhere?  It probably doesn't belong in the description of `specialize`.  I think this should only take a few sentences.

> Well, evaluation and typing are two very different things. If you want to write the documentation you should start by never naming one by the other imho. Otherwise you will confuse the reader.

Did you mean to be snarky here?  I described exactly what I meant by "evaluation" for the purpose of our discussion, which is standard terminology for software (though not for Coq), which you quoted.

Reply:
This discussion is going nowhere sorry. I probably don't understand what you are trying to say. Last attempt:
`H` is a variable, it takes place of something else. And it can be used as the thing it replaces. Examples:
- In a term like `fun H:nat => H + 1`, `H` is a variable of type `nat` and therefore we can use it as a `nat` (like adding `1` to it).
- In a term like `fun H:(nat -> nat) => (H 1)`, `H` is a variable of type `nat -> nat` and therefore we can use it as a function (since its type is a product), like applying it to `1`.
- In a term like `fun H:(forall n m : nat, n >= m), (H 0 1)`,  `H` is a variable of type `forall n m : nat, n >= m` and therefore we can use it as a function (since its type is a (dependent) product), like applying it to `0` and `1`.

So yes we can say that "`H` is not a function" in the last example as we can say that "`H` is not a natural" in the first example, but it is a variable with a functional type so yes of course we can use it as a function. That's precisely what variables are meant to be used, I really don't see your point here.

Reply:
Thanks for trying to explain.  Some discussions don't work very well by email.  Perhaps I can figure this out with a little help from @Zimmi48 when he has some time.

A last (full) restatement of my question:

Consider

```
Goal (forall n m, n >= m) -> True.
intros.                             (* H : forall n m : nat, n >= m *)
let x := type of (H 1) in idtac x.  (* (forall m : nat, 1 >= m) *)
```
 
`(H 1)` represents the application of a function.  However `H` is not a function, merely a variable with a known type. The type of `H` (`forall n m : nat, n >= m`), which doesn't look like a function, either.  So how can this term be considered application of a function?  The doc for function application [here](https://coq.inria.fr/distrib/current/refman/language/core/assumptions.html#function-application) says the first term (`H`) must be a function.  I thought the answer might be something like:
- the description of function application  is incorrect/incomplete, OR
- The processing of `(H 1)` looks at the type of `H` to infer the type of `(H 1)`, which, if correct, ought to be mentioned in the doc.

(In contrast, Nat.add is function (right?), so `(Nat.add 1)` fits the documentation for function application.  So seems like there are 2 cases to consider for function application.)

Reply:
Although the documentation describes how a function is supposed to be built, it does not say that only functions can be applied: syntax of application is [here](https://coq.inria.fr/distrib/current/refman/language/core/assumptions.html#grammar-token-term_application). The form is `term1 arg`, where [term1](https://coq.inria.fr/distrib/current/refman/language/core/basic.html#grammar-token-term1) is any kind of term/evar/etc. So really the only restriction of application is typing.

Reply:
Well, that doc is under the heading **Function application** and it says things like
![image](https://user-images.githubusercontent.com/1253341/205215583-cdffde48-178e-4ebc-80f2-24c5548c2660.png)
I read that as requiring a function.  But if what you say is correct, I will revise the description.  Could you give an example in which `term1` is not a function?

I expect that `(H 1)` always gives an error if the type of `H` is not a `forall ...`.  Counterexamples welcome.  (I like being proved wrong, it means I'm learning something.)

Reply:
Yes, that is right: the only terms that can be applied are those who have a product type. Whatever the term is (a hypothesis (i.e. a variable), an evar, a complex term, a lemma etc...).
Your initial question was "why can we apply `H` to something since `H` (of type `forall bla`) is a hypothesis and not a function". The answer is: yes it is a function since it has a product type.

Reply:
> ![image](https://user-images.githubusercontent.com/1253341/205215583-cdffde48-178e-4ebc-80f2-24c5548c2660.png)

Remark: I don't understand the meaning of the subscript `fun` in the documentation. It seems to refer to `term1` (when I click on it it takes me to the `term1` entry, so I guess it is just a way to distinguish this term from the other `term1` of the expression. Is this a global convention? Any subscrip added to a grammatical entry is actually just a tag for easy reference?

Reply:
> Any subscript added to a grammatical entry is actually just a tag for easy reference?

Exactly.

Reply:
The current text suggests that you can specialize hypotheses such `H : A -> B`, but I wasn't able to construct a working example.  If is indeed possible, could you provide a trivial example?

This is that last thing I tried:

```
Goal forall A B:Prop, (1=1) -> (A -> B) -> True.
intros A B H1 H.
(*
  A, B : Prop
  H1 : 1 = 1
  H : A -> B
*)
specialize (H H1).
(*
  In environment
  A, B : Prop
  H1 : 1 = 1
  H : A -> B
  The term "H1" has type "1 = 1" while it is expected to have type "A".
Raised at file "pretyping/coercion.ml", line 529,
```

Reply:
This is a simple type error:
- `H` is of type `A -> B` so you need to apply it to something of type `A`.
- you apply it to `H1`, which does not have type `A`, type error.
- If you had something of type A, then you could apply `H` to it:

Example:
```coq
Goal forall A B:Prop, (A -> B) -> A -> True.
Proof.
  intros A B H prfA. 
  specialize (H prfA).
```

Reply:
Thanks.  That seems a more intuitive way to do forward reasoning in hypotheses than `apply`.  `specialize` changes `H` while `apply` changes `prfA`.  Maybe worth mentioning as a tip in the doc.

For example:

```
Goal forall A B:Prop, (A -> B) -> A -> True.
Proof.
  intros A B H prfA.
  apply H in prfA.
(*
  1 subgoal
  A, B : Prop
  H : A -> B
  prfA : B
*)
```

versus

```
1 subgoal
A, B : Prop
H : B
prfA : A
```

`apply` may create new goals (and `specialize` can only apply premises in order):

![image](https://user-images.githubusercontent.com/1253341/205519170-5b5e5fa0-8a8a-417e-83ce-c3d962c75d82.png)

Reply:
> I expect that `(H 1)` always gives an error if the type of `H` is not a `forall ...`. Counterexamples welcome. (I like being proved wrong, it means I'm learning something.)

Not germane to the discussion here, but there are three technical counterexamples: the type of `H` may be reducible to a product without being syntactically a product; the type of `H` may be an evar which then is instantiated with a product during unification; or the type of `H` may be a type with a registered coercion to `Funclass`.

Reply:
Also, is "function" defined anywhere in the documentation?  It seems like a lot of the confusion here is about what, exactly, counts as a "function".  Is it something whose type is `forall ...`?  Is it something which reduces to `fun ...` or `fix ...` or `cofix ...` or ...?  I'd go with the former, even though it takes some adjustment to see `Axiom DNE : forall P, ~~P -> P` as making `DNE` a function.

Reply:
@JasonGross Regarding your first three comments, I now have "If the head constant of _one_term_ has the type **forall ...**" and then give (H 1) as an example of function application.  I think makes it unnecessary to mention your corner cases here.  Look at the what I wrote (after I push my changes again).

I added a todo about your last comment in `assumptions.rst`.  I want to wrap up the tactics chapter.